src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/ThrowHelper.cs

```diff
@@ -1,26 +1,97 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using Microsoft.Quic;
+using System.Security.Authentication;
+using static Microsoft.Quic.MsQuic;
+
 namespace System.Net.Quic.Implementations.MsQuic
 {
     internal static class ThrowHelper
     {
-        internal static Exception GetConnectionAbortedException(long errorCode)
+        internal static QuicException GetConnectionAbortedException(long errorCode)
         {
             return errorCode switch
             {
-                -1 => new QuicOperationAbortedException(), // Shutdown initiated by us.
-                long err => new QuicConnectionAbortedException(err) // Shutdown initiated by peer.
+                -1 => GetOperationAbortedException(), // Shutdown initiated by us.
+                long err => new QuicException(QuicError.ConnectionAborted, SR.Format(SR.net_quic_connectionaborted, err), err, null) // Shutdown initiated by peer.
             };
         }
 
-        internal static Exception GetStreamAbortedException(long errorCode)
+        internal static QuicException GetStreamAbortedException(long errorCode)
         {
             return errorCode switch
             {
-                -1 => new QuicOperationAbortedException(), // Shutdown initiated by us.
-                long err => new QuicStreamAbortedException(err) // Shutdown initiated by peer.
+                -1 => GetOperationAbortedException(), // Shutdown initiated by us.
+                long err => new QuicException(QuicError.StreamAborted, SR.Format(SR.net_quic_streamaborted, err), err, null) // Shutdown initiated by peer.
             };
         }
+
+        internal static QuicException GetOperationAbortedException(string? message = null)
+        {
+            return new QuicException(QuicError.OperationAborted, message ?? SR.net_quic_operationaborted, null, null);
+        }
+
+        internal static Exception GetExceptionForMsQuicStatus(int status, string? message = null)
+        {
+            //
+            // Start by checking for statuses mapped to QuicError enum
+            //
+
+            if (status == QUIC_STATUS_ADDRESS_IN_USE) return new QuicException(QuicError.AddressInUse, SR.net_quic_address_in_use, null, null);
+            if (status == QUIC_STATUS_UNREACHABLE) return new QuicException(QuicError.HostUnreachable, SR.net_quic_host_unreachable, null, null);
```

Comment:
So the message from the input parameter gets ignored? Should we append or rather prepend to the generic `SR.` message?

Reply:
I don't think the string in the message parameter would add anything useful in cases when we have a `QuicError` member. It is still used e.g. for `InternalError`

Reply:
The messages passed to the helper contain information which method failed, e.g.: "ListenerOpen failed." I guess that the operation that failed would be possible to figure out from the stack trace so it might be redundant. Yet, we'll have it some cases not in others...

Reply:
so, let's consider that a call to `ListenerOpen` fails with `QUIC_STATUS_ADDRESS_IN_USE`, If we were to include the message passed to the helper, we end up with something like

ListenerOpen failed: The local address is already in use.

I see following problems with this:
- The start of the message is not localizable (while second part is from localizable string resources) - I don't personally care, but it looks bad
- the ListenerOpen part is confusing to the user, as he called `QuicListener.Listen` method and is not (and should not need to be) aware of any `ListenerOpen` underneath, it is an implementation leak which does not add much useful diagnostic info

The only cases where I see those messages being useful is for the "unexpected" errors which we map to `QuicError.InternalError`. If for some reason we hit an `QUIC_STATUS_INVALID_PARAMETER`, `QUIC_STATUS_INVALID_STATE`, or `QUIC_STATUS_NOT_SUPPORTED`, or something else, then the additional text may be actually useful for narrowing down which call failed, and ends up in the message of the inner exception.

Also note that we don't have this custom message available for most low-level network failures, because we receive the status in the "shutdown by transport" event.